README
See instead:
- http://visl.sdu.dk/cg3.html
- http://visl.sdu.dk/cg3/chunked/
- manual/


ChangeLog
2020-01-03  tino

	* [r13887] src/GrammarApplicator_runRules.cpp: Optimize Restore
	* [r13886] manual/contexts.xml, manual/rules.xml,
	  src/ApertiumApplicator.cpp, src/Cohort.hpp,
	  src/ContextualTest.hpp, src/GrammarApplicator.cpp,
	  src/GrammarApplicator_matchSet.cpp,
	  src/GrammarApplicator_runContextualTest.cpp,
	  src/GrammarApplicator_runRules.cpp, src/GrammarWriter.cpp,
	  src/NicelineApplicator.cpp, src/Rule.hpp, src/Strings.cpp,
	  src/Strings.hpp, src/TextualParser.cpp, test/T_IgnoreRestore,
	  test/T_IgnoreRestore/args.txt, test/T_IgnoreRestore/expected.txt,
	  test/T_IgnoreRestore/grammar.cg3,
	  test/T_IgnoreRestore/grammar.cg3b.10043,
	  test/T_IgnoreRestore/input.txt: Implement rule type Restore, rule
	  flag Ignored and LookIgnored, and contextual option 'I' (closes
	  #51)

2019-12-30  tino

	* [r13878] src/GrammarApplicator.cpp: Fix segfault when varstring
	  relations result in new tags

2019-12-12  tino

	* [r13858] src/Grammar.cpp, src/GrammarApplicator.cpp, src/Tag.cpp,
	  src/inlines.hpp, src/parser_helpers.hpp,
	  test/T_SplitCohort/grammar.cg3: Fix //r regex tags implying but
	  not actually implying ^$

2019-11-11  tino

	* [r13756] src/GrammarApplicator.cpp, src/GrammarApplicator.hpp,
	  src/NicelineApplicator.cpp, src/main.cpp, src/options.hpp:
	  Implement --dep-absolute

2019-11-04  tino

	* [r13723] src/GrammarApplicator_runRules.cpp,
	  test/T_RemCohort/expected.txt: When removing cohorts, append them
	  to the previous cohort's enclosed cohorts, if any. This preserves
	  the input order much better.

2019-10-30  tino

	* [r13706] src/version.hpp: Release 1.3.1

2019-10-25  tino

	* [r13701] src/GrammarApplicator_runContextualTest.cpp,
	  test/T_Omniscan/grammar.cg3: Shorthand @< and @> should fail when
	  prev/next missing (fixes #47)

2019-10-13  tino

	* [r13693] python/CMakeLists.txt, python/setup.py.in: Python module
	  cg3 alias of constraint_grammar

2019-09-29  tino

	* [r13684] ChangeLog, newsletters/2019-09-29.txt, src/version.hpp:
	  Release 1.3.0

2019-09-24  tino

	* [r13676] python/CMakeLists.txt: No, SWIG >= 3.0 is required due
	  to namespaces
	* [r13675] python/CMakeLists.txt: SWIG 3.0 -> 2.0

2019-09-23  tino

	* [r13674] cg3.pc.in: Missed ${prefix}/

2019-09-17  tino

	* [r13661] src/TextualParser.cpp: Forward in-template status to
	  linked tests

2019-09-16  tino

	* [r13659] CMakeLists.txt, src/CMakeLists.txt: Use GNUInstallDirs
	  and relative paths

2019-09-14  tino

	* [r13658] python/CMakeLists.txt: LIST_DIR -> BINARY_DIR
	* [r13657] CMakeLists.txt, python/CMakeLists.txt: Add
	  PYTHON_INSTALL_PARAMS to help with packaging

2019-09-05  tino

	* [r13606] src/AST.hpp, src/ApertiumApplicator.cpp, src/Cohort.cpp,
	  src/CohortIterator.cpp, src/ContextualTest.cpp,
	  src/FSTApplicator.cpp, src/Grammar.cpp,
	  src/GrammarApplicator.cpp, src/GrammarApplicator_matchSet.cpp,
	  src/GrammarApplicator_reflow.cpp,
	  src/GrammarApplicator_runContextualTest.cpp,
	  src/GrammarApplicator_runGrammar.cpp,
	  src/GrammarApplicator_runRules.cpp, src/GrammarWriter.cpp,
	  src/MatxinApplicator.cpp, src/MweSplitApplicator.cpp,
	  src/NicelineApplicator.cpp, src/PlaintextApplicator.cpp,
	  src/Reading.cpp, src/Rule.cpp, src/Set.cpp, src/SingleWindow.cpp,
	  src/Tag.cpp, src/TextualParser.cpp, src/TextualParser.hpp,
	  src/Window.cpp, src/cg_conv.cpp, src/cg_proc.cpp,
	  src/flat_unordered_map.hpp, src/flat_unordered_set.hpp,
	  src/inlines.hpp, src/interval_vector.hpp, src/libcg3.cpp,
	  src/process.hpp, src/uextras.cpp: Don't warn about wrong ** in
	  templates (workaround); Lots of 0 -> nullptr

2019-09-02  tino

	* [r13587] src/BinaryGrammar_read.cpp,
	  src/BinaryGrammar_read_10043.cpp, src/Grammar.cpp,
	  src/GrammarWriter.cpp, src/TextualParser.cpp, src/cg-relabel.cpp,
	  src/cg_comp.cpp, src/cg_proc.cpp, src/inlines.hpp,
	  src/libcg3.cpp, src/main.cpp: Turn magic constants into functions
	* [r13586] manual/sets.xml, src/Set.cpp, src/parser_helpers.hpp,
	  test/T_Unification/expected.txt, test/T_Unification/grammar.cg3,
	  test/T_Unification/grammar.cg3b.10043: Implement multi-use
	  unification for && sets (issue #44)
	* [r13585] src/GrammarApplicator.cpp, src/GrammarApplicator.hpp,
	  src/GrammarApplicator_matchSet.cpp,
	  src/GrammarApplicator_runRules.cpp, src/inlines.hpp: Handle
	  multiple active &&sets at a time.

2019-08-30  tino

	* [r13577] CMakeLists.txt, python/CMakeLists.txt: Allow configuring
	  Python module without building or installing it

2019-08-29  tino

	* [r13560] CMakeLists.txt: Default policy to newest so that we can
	  catch deprecations; Be nicer when built as subproject; Set CPack
	  version

2019-08-26  tino

	* [r13555] src/GrammarApplicator_runRules.cpp: Move/Switch endless
	  loop no longer fatal
	* [r13554] .gitignore, .travis.yml, CMakeLists.txt, cmake.sh,
	  python, python/CMakeLists.txt, python/constraint_grammar.i.in,
	  python/setup.py.in: SWIG Python 3 bindings, by Lokendra Singh
	  <lokendras1998@gmail.com> (closes #37)
	  
	  git-svn-author: lokendras1998@gmail.com

2019-08-21  tino

	* [r13551] src/GrammarApplicator_matchSet.cpp: Don't let barriers
	  test the static reading (issue #42)

2019-08-19  tino

	* [r13550] src/GrammarApplicator_runRules.cpp: Whoops, only do that
	  for the * reading
	* [r13549] src/GrammarApplicator_runRules.cpp: If there aren't any
	  sub-readings, return the primary reading
	* [r13548] src/GrammarApplicator.hpp: Not supposed to be
	  [[noreturn]]

2019-08-17  tino

	* [r13547] src/GrammarApplicator_runRules.cpp,
	  test/T_Append/expected.txt, test/T_Append/grammar.cg3,
	  test/T_Append/input.txt: Copy Except will now except from
	  sub-readings as well as main reading

2019-08-13  tino

	* [r13542] src/GrammarApplicator_reflow.cpp, src/Strings.cpp,
	  src/Strings.hpp: Convert %[UuLl] markers to control codes to
	  avoid having combined % accidentally match %L

2019-08-09  tino

	* [r13538] src/TextualParser.cpp, src/TextualParser.hpp: More

src/Rule.h
/*
* Copyright (C) 2007-2013, GrammarSoft ApS
* Developed by Tino Didriksen <mail@tinodidriksen.com>
* Design by Eckhard Bick <eckhard.bick@mail.dk>, Tino Didriksen <mail@tinodidriksen.com>
*
* This file is part of VISL CG-3
*
* VISL CG-3 is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* VISL CG-3 is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with VISL CG-3.  If not, see <http://www.gnu.org/licenses/>.
*/

#pragma once
#ifndef c6d28b7452ec699b_RULE_H
#define c6d28b7452ec699b_RULE_H

#include "stdafx.h"
#include "Strings.h"
#include "Tag.h"
#include "Cohort.h"
#include "ContextualTest.h"

namespace CG3 {

	class Grammar;
	class Set;

	// This must be kept in lock-step with Strings.h's FLAGS
	enum {
		RF_NEAREST      = (1 <<  0),
		RF_ALLOWLOOP    = (1 <<  1),
		RF_DELAYED      = (1 <<  2),
		RF_IMMEDIATE    = (1 <<  3),
		RF_LOOKDELETED  = (1 <<  4),
		RF_LOOKDELAYED  = (1 <<  5),
		RF_UNSAFE       = (1 <<  6),
		RF_SAFE         = (1 <<  7),
		RF_REMEMBERX    = (1 <<  8),
		RF_RESETX       = (1 <<  9),
		RF_KEEPORDER    = (1 << 10),
		RF_VARYORDER    = (1 << 11),
		RF_ENCL_INNER   = (1 << 12),
		RF_ENCL_OUTER   = (1 << 13),
		RF_ENCL_FINAL   = (1 << 14),
		RF_ENCL_ANY     = (1 << 15),
		RF_ALLOWCROSS   = (1 << 16),
		RF_WITHCHILD    = (1 << 17),
		RF_NOCHILD      = (1 << 18),
		RF_ITERATE      = (1 << 19),
		RF_NOITERATE    = (1 << 20),
		RF_UNMAPLAST    = (1 << 21),
		RF_REVERSE      = (1 << 22),
		RF_SUB          = (1 << 23),
	};

	class Rule {
	public:
		UChar *name;
		uint32_t wordform;
		uint32_t target;
		uint32_t childset1, childset2;
		uint32_t line, number;
		uint32_t varname, varvalue; // ToDo: varvalue is unused
		uint32_t flags;
		int32_t section;
		int32_t sub_reading;
		// ToDo: Add proper "quality" quantifier based on num_fail, num_match, total_time
		double weight, quality;
		KEYWORDS type;
		Set *maplist;
		Set *sublist;
		
		mutable ContextList tests;
		mutable ContextList dep_tests;
		mutable uint32_t num_fail, num_match;
		mutable double total_time;
		mutable ContextualTest *dep_target;

		Rule();
		~Rule();
		void setName(const UChar *to);
		
		void resetStatistics();

		void addContextualTest(ContextualTest *to, ContextList& head);
		void reverseContextualTests();

		static bool cmp_quality(const Rule *a, const Rule *b);

		static inline size_t cmp_hash(const Rule* r) {
			return hash_sdbm_uint32_t(r->number);
		}
		static inline bool cmp_compare(const Rule* a, const Rule* b) {
			return a->number < b->number;
		}
	};

	struct compare_Rule {
		static const size_t bucket_size = 4;
		static const size_t min_buckets = 8;

		inline size_t operator() (const Rule* r) const {
			return Rule::cmp_hash(r);
		}

		inline bool operator() (const Rule* a, const Rule* b) const {
			return Rule::cmp_compare(a, b);
		}
	};

	typedef std::vector<Rule*> RuleVector;
	typedef std::map<uint32_t,Rule*> RuleByLineMap;
	typedef stdext::hash_map<uint32_t,Rule*> RuleByLineHashMap;
	typedef stdext::hash_map<const Rule*, CohortSet, compare_Rule> RuleToCohortsMap;
}

#ifdef __GNUC__
#ifndef HAVE_BOOST
#if GCC_VERSION < 40300
namespace __gnu_cxx {
	template<> struct hash< CG3::Rule* > {
		size_t operator()( const CG3::Rule *x ) const {
			return CG3::Rule::cmp_hash(x);
		}
	};
}
#endif
#endif
#endif

#endif
